File: python/viz/plot_full_spring.py
import numpy as np
import matplotlib.pyplot as plt
from viz.spring import Spring

def plot_full_spring(x1, y1, x2, y2, R, RATIO, nc, color='k', LW=1, ax=None):
    """
    Plot full spring matching MATLAB PlotFullSpring.m
    
    Parameters
    ----------
    x1, y1 : float
        Start point coordinates
    x2, y2 : float
        End point coordinates
    R : float
        Spring radius
    RATIO : float
        Spring length ratio
    nc : int
        Number of coils
    color : str or array_like, optional
        Color specification, by default 'k'
    LW : float, optional
        Line width, by default 1
    ax : matplotlib.axes.Axes, optional
        Axes to plot on, by default None (creates new figure)
    """
    
    if ax is None:
        fig, ax = plt.subplots()
    
    # Create an instance of spring
    spr = Spring(R, nc)
    
    # Direction vector
    n = np.array([x2, y2]) - np.array([x1, y1])
    
    # Coordinates
    p1 = np.array([x1, y1])
    p2 = np.array([x1 + 0.5 * (1-RATIO) * n[0], y1 + 0.5 * (1-RATIO) * n[1]])
    p3 = np.array([x2 - 0.5 * (1-RATIO) * n[0], y2 - 0.5 * (1-RATIO) * n[1]])
    p4 = np.array([x2, y2])
    
    # Get coordinates of spring
    x_s, y_s = spr.get_spr(p2, p3)
    
    # Plot spring
    ax.plot(x_s, y_s, color=color, linewidth=LW)
    
    # Plot legs
    ax.plot([p1[0], p2[0]], [p1[1], p2[1]], color=color, linewidth=LW)
    ax.plot([p3[0], p4[0]], [p3[1], p4[1]], color=color, linewidth=LW)

File: matlab/PlotFullSpring.m
function [] = PlotFullSpring(x1, y1, x2, y2, R, RATIO, nc, color, LW)

% Check input arguments
if nargin < 7; nc = 5; end
if nargin < 8; color = 'k'; end
if nargin < 9; LW = 1; end

% Create an instance of spring
spr = Spring(R, nc);

% Length
% L = sqrt((x2-x1)^2 + (y2-y1)^2);

% Spring length
% Ls = RATIO * L;

% Leg length
% Ll = (L - Ls)/2;

% Direction vector
n = [x2; y2] - [x1; y1];

% Coordinates
p1 = [x1; y1];
p2 = [x1 + 0.5 * (1-RATIO) * n(1); y1 + 0.5 * (1-RATIO) * n(2)];
p3 = [x2 - 0.5 * (1-RATIO) * n(1); y2 - 0.5 * (1-RATIO) * n(2)];
p4 = [x2; y2];

% Get coordinates of spring
[x_s, y_s] = spr.getSpr(p2, p3);

% Plot spring
plot(x_s, y_s, 'Color', color, 'LineWidth', LW);

% Plot legs
plot([p1(1); p2(1)], [p1(2); p2(2)], 'Color', color, 'LineWidth', LW);
plot([p3(1); p4(1)], [p3(2); p4(2)], 'Color', color, 'LineWidth', LW);
